Microsoft.DotNet.Interactive.PowerShell/PowerShellKernel.cs

```diff
@@ -0,0 +1,311 @@
+                        }
+                    }
+                }
+                catch (Exception e)
+                {
+                    exception = e;
+                }
+                finally
+                {
+                    _runspaceSemaphore.Release();
+                }
+            }
+
+            if (!cancellationSource.IsCancellationRequested)
+            {
+                if (exception != null)
+                {
+                    string message = null;
+                    context.Publish(new CommandFailed(exception, submitCode, message));
+                }
+            }
+            else
+            {
+                context.Publish(new CommandFailed(null, submitCode, "Command cancelled"));
+            }
+        }
+
+        private async Task HandleRequestCompletion(
+            RequestCompletion requestCompletion,
+            KernelInvocationContext context)
+        {
+            var completionRequestReceived = new CompletionRequestReceived(requestCompletion);
+
+            context.Publish(completionRequestReceived);
+
+            var completionList =
+                await GetCompletionList(
+                    requestCompletion.Code, 
+                    requestCompletion.CursorPosition);
+
+            context.Publish(new CompletionRequestCompleted(completionList, requestCompletion));
+        }
+
+        private Task HandleCancelCurrentCommand(
+            CancelCurrentCommand cancelCurrentCommand,
+            KernelInvocationContext context)
+        {
+            var reply = new CurrentCommandCancelled(cancelCurrentCommand);
+            lock (_cancellationSourceLock)
+            {
+                _cancellationSource.Cancel();
+                _cancellationSource = new CancellationTokenSource();
+            }
+
+            context.Publish(reply);
+
+            return Task.CompletedTask;
+        }
+
+        #endregion
+
+        public bool IsCompleteSubmission(string code)
+        {
+            Parser.ParseInput(code, out Token[] tokens, out ParseError[] errors);
+            return errors.Length > 0;
+        }
+
+        private void PublishOutput(
+            string output,
+            KernelInvocationContext context,
+            IKernelCommand command)
+        {
+            var formattedValues = new List<FormattedValue>
+                        {
+                            new FormattedValue(
+                                PlainTextFormatter.MimeType, output)
+                        };
+
+            context.Publish(
+                new StandardOutputValueProduced(
+                    output,
+                    command,
+                    formattedValues));
+        }
+
+        private void PublishStreamRecord(
+            object output,
+            KernelInvocationContext context,
+            IKernelCommand command)
+        {
+            context.Publish(
```

Comment:
We added a `KernelInvocationContext.DisplayAsync` method that might simplify this for you.